gravitee-plugin-core/src/main/java/io/gravitee/plugin/core/internal/PluginRegistryImpl.java

```diff
@@ -102,99 +96,101 @@ protected void doStart() throws Exception {
+            // As plugin arrive sorted by reverse file date
+            // we can exclude duplicate
+            .distinct()
+            .doOnNext(plugin -> {
+                plugins.add(plugin);
+                eventManager.publishEvent(PluginEvent.DEPLOYED, plugin);
+            })
+            .doOnComplete(() -> {
+                printPlugins();
+                eventManager.publishEvent(PluginEvent.ENDED, null);
+            })
+            .ignoreElements()
+            .blockingAwait();
     }
 
-    private void loadPluginsFromRegistry(String registryPath) throws Exception {
-        File registryDir = new File(registryPath);
+    private Flowable<PluginImpl> loadPluginsFromPath(final String pluginPathAsString) throws IOException {
+        File pluginDir = new File(pluginPathAsString);
 
-        // Quick sanity check on the install root
-        if (!registryDir.isDirectory()) {
-            LOGGER.error("Invalid registry directory, {} is not a directory.", registryDir.getAbsolutePath());
-            throw new RuntimeException("Invalid registry directory. Not a directory: " + registryDir.getAbsolutePath());
+        // Quick sanity check
+        if (!pluginDir.isDirectory()) {
+            return Flowable.error(
+                new IllegalArgumentException("Invalid registry directory. Not a directory: " + pluginDir.getAbsolutePath())
+            );
         }
 
-        loadPlugins(registryDir);
-    }
-
-    private void loadPlugins(File registryDir) throws Exception {
-        Path registryPath = registryDir.toPath();
-        LOGGER.info("Loading plugins from {}", registryDir);
-
-        try (DirectoryStream<Path> stream = FileUtils.newDirectoryStream(registryPath, ZIP_GLOB)) {
-            List<CompletableFuture<?>> futures = new ArrayList<>();
-            Iterator<Path> archiveIte = stream.iterator();
-
-            if (archiveIte.hasNext()) {
-                while (archiveIte.hasNext()) {
-                    final Path path = archiveIte.next();
-                    futures.add(runAsync(() -> loadPlugin(registryDir, path), executor));
-                }
-            } else {
-                LOGGER.warn("No plugin has been found in {}", registryDir);
-            }
-
-            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).get();
-
-            init = true;
-        } catch (IOException ioe) {
-            LOGGER.error("An unexpected error occurs", ioe);
-            throw ioe;
-        }
+        final Path pluginPath = pluginDir.toPath();
+        log.info("Loading plugins from {}", pluginDir);
+
+        final DirectoryStream<Path> stream = FileUtils.newDirectoryStream(pluginPath, ZIP_GLOB);
+        return Flowable
+            .fromIterable(stream)
+            .subscribeOn(Schedulers.from(executor))
+            .map(path -> loadPlugin(pluginDir, path))
+            .filter(PluginImpl::valid)
+            .doOnComplete(() -> this.init = true)
```

Comment:
Shouldn't you set the init value, on the main chain when all plugin are treated line 135 for example with a DoOnComplete after the ignore Element or event after the blockingAwait?

Reply:
It was done like this before... but I wondered too. I suppose it is more something like "in progress" so any new caller would have an error

Reply:
It is set as soon as possible